### PR TITLE
cpu/nrf52 i2c: Always buffer writes

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

The underlying peripheral can only read from RAM. This uses the existing infrastructure (already needed to work around the lack of a hardware support for I2C_NOSTART) to unconditionally copy any to-be-sent data into RAM.

There is certainly an alternative where we determine at runtime whether the data is in RAM or in ROM and only lock and copy if it's in ROM, but I'd only do that if we have concrete benchmarks that show that this is an issue.

### Testing procedure

* To see that the old behavior works, run the saul demo on a microbit-v2, or any other board's I2C.
* To see that things work even from flash, run the saul demo on a microbit-v2 with my lsm303agr-0.3 branch -- that updated the driver such that it can send its commands from ROM, which broke things here.

### CODE REVIEW

File: cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c [hunk 1]
```diff
@@ -331,6 +339,27 @@ int i2c_read_regs(i2c_t dev, uint16_t addr, uint16_t reg,
 
 int i2c_write_bytes(i2c_t dev, uint16_t addr, const void *data, size_t len,
                     uint8_t flags)
+{
+    /* These are critical for the memcpy; direct_i2c_write_bytes makes some
```
Comment (benpicco) [inline cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: How about a fast path

```C
if ((uintptr_t)data >= RAM_START_ADDR) {
    return direct_i2c_write_bytes(addr, data, len, flags);
}
```

Comment (chrysn) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: That'd be relying on an [undocumented](https://duckduckgo.com/?q=%22RAM_START_ADDR%22+site%3Adoc.riot-os.org&kt=h&ia=web) constant.

On the nRF52840, both code and XPI (memory mapped external QSPI) are all below RAM_START_ADDR, but the nRF9160 [memory map](https://infocenter.nordicsemi.com/topic/ps_nrf9160/memory.html) has a lot of stuff below; is there a RAM_END_ADDR too?

Comment (dylad) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: No but you have RAM_LEN IIRC.

Comment (chrysn) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: None of these symbols even exist. What I got to work is

```patch
+    extern char *_ram_start_addr;
+    extern unsigned int _ram_length;
+    if ((char*)data >= _ram_start_addr && (char*)data < _ram_start_addr + _ram_length) {
+        return direct_i2c_write_bytes(dev, addr, data, len, flags);
+    }
+
```

-- that is reaching more deeply into linker stuff that I'm usually comfortable doing, but if it's what you two prefer, so be it.

Comment (benpicco) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: > None of these symbols even exist.

huh, we have 

```
cpu/nrf5x_common/Makefile.include
4:RAM_START_ADDR ?= 0x20000000
```
(actually could move that to `cpu/cortexm_common`) and 

```
makefiles/cflags.inc.mk
124:ifneq (,$(RAM_START_ADDR))
125:  CFLAGS += -DCPU_RAM_BASE=$(RAM_START_ADDR)
```

Comment (chrysn) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:347]: Thanks for the symbol name, that indeed works without externs.

Up to today it was unlikely that someone would send data from flash, and that that someone just spent over a day hunting down three different subtle bugs that all just started showing when an innocent change in a driver that was already well tested caused trouble. I don't wish that same trouble on the first person who uses nRF9610 to send valid data from one place to another.

I've updated this PR to do the two-sided check, please review.

**Review by benpicco — APPROVED** [review]

**Comment by maribu** [comment]
The branch name is quite confusing :D

**Comment by chrysn** [comment]
If that's confusing than please don't look at #20299, which is `nrf52-spi-bugs-2`. Maybe it's a biotech thing: Genes are named after the property an organism has when you knock out the gene, so here the branch is named after the property that is removed from RIOT when it is applied :-)

File: cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c [hunk 2]
```diff
@@ -331,6 +339,31 @@ int i2c_read_regs(i2c_t dev, uint16_t addr, uint16_t reg,
 
 int i2c_write_bytes(i2c_t dev, uint16_t addr, const void *data, size_t len,
                     uint8_t flags)
+{
+    if ((unsigned int)data >= CPU_RAM_BASE && (unsigned int)data < CPU_RAM_BASE + CPU_RAM_SIZE) {
```
Comment (kaspar030) [inline cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: This is done a bit simpler for spi: https://github.com/RIOT-OS/RIOT/blob/1073df2ce2804e3290467cba7ca12e36b0d557d5/cpu/nrf5x_common/periph/spi_nrf52_nrf9160.c#L60

Would that work here, too? Is that to relaxed for SPI?

Comment (kaspar030) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: ... embassy-nrf does this:

```
const SRAM_LOWER: usize = 0x2000_0000;
const SRAM_UPPER: usize = 0x3000_0000;
ptr >= SRAM_LOWER && (ptr + len * core::mem::size_of::<T>()) < SRAM_UPPER

Comment (chrysn) [reply cpu/nrf5x_common/periph/i2c_nrf52_nrf9160.c:343]: The SPI version is probably good enough, we could unify. The embassy version is IMO excessive, as the slice is a single object, and objects don't span memmapping areas.

**Pull request merged**